core/src/main/scala/org/apache/spark/util/SizeEstimator.scala

```diff
@@ -221,8 +226,13 @@ object SizeEstimator extends Logging {
         case _ =>
           val classInfo = getClassInfo(cls)
           state.size += alignSize(classInfo.shellSize)
-          for (field <- classInfo.pointerFields) {
-            state.enqueue(field.get(obj))
+          // avoid an iterator based for loop for performance
+          var index = 0
+          val fieldCount = classInfo.fieldOffsets.length
+          val us = Unsafe.instance
+          while (index < fieldCount) {
+            state.enqueue(us.getObject(obj, classInfo.fieldOffsets(index).toLong))
```

Comment:
I understand avoiding reflection, but this is a dicier way to access fields of an object. I don't have a specific reason this would fail but the fact that it uses unsafe is riskier. Is this worth it?